docs: add how to send delay event doc (#50)
linkall-labs/docs: user-manual/how-to/producing/send-delay-event.md

category: 6321d3a56093c3010397e4aa
---

> # Send delay events to Vanus with [CloudEvents](https://github.com/cloudevents) SDK

Reviewer: ```suggestion
# Send an event with delay using the CloudEvents SDK
```

> The following document will teach you how to build a simple producer to send delay events to Vanus.

Reviewer: ```suggestion
he following document will teach you how to build a simple producer with a timer to delay an Event. 
```

## Operations

**Prerequisites**

To send events to Vanus, you must meet the following prerequisites:
1. Have a running [Vanus](https://github.com/linkall-labs/docs/blob/main/user-manual/getting-started/install/k8s(recommended).md) cluster.
2. Have [vsctl](https://github.com/linkall-labs/docs/blob/main/user-manual/how-to/vsctl.md).

> 3. Have created an [Evevtbus](https://github.com/linkall-labs/docs/blob/main/concepts/eventbus.md) named quick-start.

Reviewer: ```suggestion
3. Have created an [Eventbus](https://github.com/linkall-labs/docs/blob/main/concepts/eventbus.md) named quick-start.
```

	xceVanusDeliveryTime = "xvanusdeliverytime"
)

var (
	endpoint = os.Getenv("VANUS_GATEWAY")
)

func main() {
	// step 1: build a CloudEvents client
	client, err := ce.NewClientHTTP()
	if err != nil {
		panic(fmt.Sprintf("new cloudevents client failed, err: %s\n", err.Error()))
	}

	var target string
	if strings.HasPrefix(endpoint, httpPrefix) {
		target = fmt.Sprintf("%s/gateway/%s", endpoint, eventbus)
	} else {
		target = fmt.Sprintf("%s%s/gateway/%s", httpPrefix, endpoint, eventbus)
	}

	// step 2: specify the target
	ctx := ce.ContextWithTarget(context.Background(), target)
	// step 3: build a CloudEvent
	event := ce.NewEvent()
	event.SetID(uuid.NewString())
	event.SetSource("event-source")
	event.SetType("event-type")
	event.SetData(ce.ApplicationJSON, map[string]string{"hello": "world"})
	// set delivery time of delay event
	timeOfRFC3339Nano := time.Now().Add(10 * time.Second).Format(time.RFC3339Nano)
	event.SetExtension(xceVanusDeliveryTime, timeOfRFC3339Nano)

	// step 4: send the CloudEvent
	if result := client.Send(ctx, event); ce.IsUndelivered(result) {
		panic(fmt.Sprintf("failed to send event, err: %s\n", result.Error()))
	}
}
```

**Expected results**

> By running the following command after 10s, we can ensure that the simple producer successfully sent the event to Vanus Eventbus.

Reviewer: ```suggestion
By running the following command after 10s, we can guarantee that The producer successfully sent the event to the Eventbus.
```
